### ISSUE TITLE
contiguous numbering sequence

### ISSUE THREAD

**Issue by giulia-berto** [brainlife/validator-neuro-wmc#3]
I have some comments on this: `"Because no name entries may be blank, and because all entities in the .names field must have at least one exemplar streamline (open to debate), the numbering sequence used in the .index vector should be contiguously sequenced (i.e. 1:1:n, no skipping numbers)"`. Why is it so important that the numbering sequence should be contiguous? Let's say for example that I'd like to segment 3 tracts, but no streamlines of the tractogram were found associated with the second tract. I think it would be better to keep the indexes 1 and 3 rather than 1 and 2. Moreover, where it is written 1:1:n, I guess n refers to the number of classes. If so, consider changing it since previously it was referring to the number of total streamlines.

**Comment by DanNBullock** [comment]
In the case you describe, you would have no instances of 2 in the index field, and there would be no need to actually list the unused name, be cause there were no streamlines of that type.  I know that may seem strange for standard segmentations, but there are other segmentation apps (i.e. https://brainlife.io/app/5dc9d8ee5017f0bde0519db5) where the number of potentially segment-able is combinatorial (n choose two).  Including all those categories where it was theoretically possible to obtain a streamline but none were, would result in a completely unwieldy data structure.

One of the overarching problems is that, in practice, a lot of apps which use WMC as an input tend to be brittle against empty categories (e.g. they will try and find a tract or a quantity and error out when one can be found).  Conceivably app creators could be tasked with making their apps robust against this.  I still worry that a number of edge cases could crop up if this were a feature of the classification structure.

Of course, none of this is set in stone, and we could change this requirement:

> all entities in the .names field must have at least one exemplar streamline (open to debate)

but we would need to have a broader discussion of what the potential impact would be.

As for your comment about the use of n as a variable, I'll update that

**Comment by soichih** [comment]
I think we should allow un-used names (name with 0 streamlines of that
type). The reason for this is that, 1) it makes it harder to do group
analysis if every subject has a different set of names and
classification indexes. and 2) "a lot of apps which use WMC as an input
tend to be brittle against empty categories" is, in my opinion something
that every App should be able to handle on its own.

I also think that, every App should have unique (and static) list of names
/ segmentation indices used, and this list should not contain any
repeated names, or names that are never segmented by the App.

Soichi

**Comment by DanNBullock** [comment]
How should the combinatorial / https://brainlife.io/app/5dc9d8ee5017f0bde0519db5  case be handled?

Also, what are we considering "segmented"  is this a binary thing (i.e. "any streams present") or is there some threshold for considering it segmented?

**Comment by soichih** [comment]
@DanNBullock Can you explain what you mean by combinatorial? 

Also, what do you mean by "handle"? What exactly do you have in mind?

**Comment by DanNBullock** [comment]
Combinatorial as in the case of n choose k (https://en.wikipedia.org/wiki/Binomial_coefficient) where the number of potential outcomes grows geometrically with the number of input options.  For example if you have 190 rois, and you segment between each of them, there are 17955 possible segmented tracts.  Many of them will be empty, but would we really want the WMC to enumerate all of them?

In the case of  https://brainlife.io/app/5dc9d8ee5017f0bde0519db5 there are many, many possible outcomes because of the segmentation is mapped, but in most cases  you won't get all of them (nor would you want to).  It isn't at all a bad thing in such a case to fail to segment most of the potential tracts with that app.  Do we really want to list all of those out every time?

Also, having a static list for an app simply isn't possible in some cases.  For example, these apps:

>  https://brainlife.io/app/5cc9ccca4b5e4502275edb56
> https://brainlife.io/app/5db737fe8aeeee161df2dc0d
> https://brainlife.io/app/5cc9e7ab4b5e4502275edb79
> (and other such apps)

Also, isn't 

> I also think that, every App should have unique (and static) list of names
/ segmentation indices used, and this list should not contain any
repeated names, or names that are never segmented by the App.

beyond the scope of this app, or is that a feature we would want to include in this app, that it looks for some check list in this validator?

**Comment by soichih** [comment]
Ah I see, yes, in such cases, those Apps should have different set of names / segmentation ID.

**Comment by DanNBullock** [comment]
> different set of names / segmentation ID.

Dynamically generated with respect to the inputs, I guess?

**Comment by soichih** [comment]
Yep. I think names should be always the same, or always different.

**Comment by DanNBullock** [comment]
What about the combinatorial / no streamline exmplar issue?

**Comment by soichih** [comment]
Such App can simply re-index the output index to make the name/index to list all the rois pairs that are tracked.

**Comment by DanNBullock** [comment]
>  re-index the output index

I'm not sure I follow what you mean by this.  Currently, my category segmentation is capable of outputting approximately 230 categories (e.g. "tracks").  That already seems  like quite a bit for the viewer to handle.  However, this 230 does not include all possible combinations of rois/categories.

For the sake of maintaining standards, should the output of the category segmentation include **all possible** categories/tracts that the (even in the event that no streamlines have been found for the category)?

**Comment by soichih** [comment]
By "re-index", I mean to update from

```
name: ['ioff.left' 'Left_TPC' 'Right_MdLF-Ang']
index: [ 0. 32. 40. 45.]
```

to this

```
name: ['ioff.left' 'Left_TPC' 'Right_MdLF-Ang']
index: [ 0. 1. 2. 3.]
```

>  That already seems like quite a bit for the viewer to handle

The current implementation of any visualizers should not influence how we standardize the wmc datatype itself. We can update the vis to handle more segmentation if necessary.

> For the sake of maintaining standards, should the output of the category segmentation include all possible categories/tracts that the (even in the event that no streamlines have been found for the category)?

Not all *possible*, but we should list all *attempted* segmentations. This way, we can separate 3 cases just by looking at the wmc datatype. 1) tract segmentation was attempted and we found streamlines 2) tract segmentation was attempted and we found 0 streamlines, 3) tract segmentation was not attempted - shown by the absence of such tract in the "names" array.
